### ISSUE TITLE
`TestParentDeletion` fails on OpenBSD

### ISSUE THREAD

**Issue by vext01** [syncthing/syncthing#5246]
This test is failing non-deterministically:
```
fremen% STRACE=model go test -v -run TestParentDeletion                                                                            
=== RUN   TestParentDeletion
2018/10/07 13:09:02.267277 set.go:80: INFO: No stored folder metadata for "default": recalculating                                 
2018/10/07 13:09:02.268209 model.go:197: INFO: Ready to synchronize "default" (default) (sendreceive)                              
2018/10/07 13:09:02.268334 model.go:1640: INFO: Device GYRZZQB-IRNPV4Z-T7TC52W-EQYJ3TT-FDQW6MW-DFLMU42-SSSU6EM-FBK2VAY client is " " named "" at
2018/10/07 13:09:02.287054 folder.go:290: INFO: Completed initial scan of sendreceive folder "default" (default)                   
--- PASS: TestParentDeletion (0.58s)
PASS
ok      github.com/syncthing/syncthing/lib/model        0.751s
fremen% STRACE=model go test -v -run TestParentDeletion
=== RUN   TestParentDeletion
2018/10/07 13:09:05.467361 set.go:80: INFO: No stored folder metadata for "default": recalculating                                 
2018/10/07 13:09:05.468368 model.go:197: INFO: Ready to synchronize "default" (default) (sendreceive)                              
2018/10/07 13:09:05.468500 model.go:1640: INFO: Device GYRZZQB-IRNPV4Z-T7TC52W-EQYJ3TT-FDQW6MW-DFLMU42-SSSU6EM-FBK2VAY client is " " named "" at
2018/10/07 13:09:05.488119 folder.go:290: INFO: Completed initial scan of sendreceive folder "default" (default)                   
--- PASS: TestParentDeletion (0.60s)
PASS
ok      github.com/syncthing/syncthing/lib/model        0.748s
fremen% STRACE=model go test -v -run TestParentDeletion
=== RUN   TestParentDeletion
2018/10/07 13:09:08.717217 set.go:80: INFO: No stored folder metadata for "default": recalculating
2018/10/07 13:09:08.718186 model.go:197: INFO: Ready to synchronize "default" (default) (sendreceive)
2018/10/07 13:09:08.718315 model.go:1640: INFO: Device GYRZZQB-IRNPV4Z-T7TC52W-EQYJ3TT-FDQW6MW-DFLMU42-SSSU6EM-FBK2VAY client is " " named "" at 
2018/10/07 13:09:08.737205 folder.go:290: INFO: Completed initial scan of sendreceive folder "default" (default)
--- FAIL: TestParentDeletion (0.09s)
    requests_test.go:589: remove /home/porter/go/src/github.com/syncthing/syncthing/lib/model/testdata/_request-402399101/foo: directory not empty
FAIL
```

Version: `f528923d1ef224a07d58fea21b0dca14ca18ba78`.

Thanks

**Comment by vext01** [comment]
I've had a a quick look at some of the failing tests for OpenBSD, and some -- like this one -- look like the calls to `removeAll` are failing because the dirs are not empty. Whatever it is, it's obviously racy too.

I don't know *anything* about go or the Syncthing codebase, but if one of the devs could tell me how to do some kind of print debugging I can look deeper. I've tried inserting `fmt.print` and `log.Printf` calls in `removeAll` but see nothing printed in the output. Am I missing something?

I'm wondering if perhaps `Glob` behaves differently on OpenBSD and the set of files that needs to be removed is incomplete... Not sure though.

**Comment by AudriusButkevicius** [comment]
l.Infoln should be available, or simple println() which should be available. RemoveAll should do a recursive remove.

**Comment by imsodin** [comment]
It may also be a good idea to check the returned errors of all those functions and fail if they are non-nil. Then it is at least clear whether that's a problem or not.

**Comment by vext01** [comment]
Alright, well `println` works.

The test creates a parent dir, then a child under the parent, removes the parent recursively and tries to scan the child.

Removing the parent sometimes fails. This is the call to ` testFs.RemoveAll(parent)` in the test, which ends up at `func (f *BasicFilesystem) RemoveAll(name string) error {` in `basicfs.go`. This is what is failing.

The removal fails because the directory is not empty, which I suppose means that something else is creating a file in the parent at the same time as the parent is being recursively removed?

Are these directories created in a non-blocking fashion? I assume `fc.addFile` is what is creating them? Maybe we need to wait for the child to be created before continuing?

Here's the relevant bit of the test
```
        parent := "foo"
        child := filepath.Join(parent, "bar")
        testFs := fs.NewFilesystem(fs.FilesystemTypeBasic, tmpDir)

        received := make(chan []protocol.FileInfo)
        fc.addFile(parent, 0777, protocol.FileInfoTypeDirectory, nil)
        fc.addFile(child, 0777, protocol.FileInfoTypeDirectory, nil)
        fc.mut.Lock()
        fc.indexFn = func(folder string, fs []protocol.FileInfo) {
                received <- fs
                return
        }
        fc.mut.Unlock()
        fc.sendIndexUpdate()

        // Get back index from initial setup
        select {
        case fs := <-received:
                if len(fs) != 2 {
                        t.Fatalf("Sent index with %d files, should be 2", len(fs))
                }
        case <-time.After(time.Second):
                t.Fatalf("timed out")
        }

        // Delete parent dir
        if err := testFs.RemoveAll(parent); err != nil {
                println("TESTFS REMOVEALL FAIL")
                t.Fatal(err)
        }
```

**Comment by imsodin** [comment]
`fc.addFile` is just "syncthetic", meaning it fakes a connection to a non-existing device - no disk io involved. Neither is the operation racy: The infos are received, the two dirs created and only then are the infos sent, which then unblocks in line 579. Meaning the removal that fails happens after pull operations. Of course there may be delay in the disk io on a lower level, but then I'd expect the removal to be delayed too.

**Comment by vext01** [comment]
Would it help if I granted one of you guys SSH access to an OpenBSD machine?

**Comment by salcedo** [comment]
http://ix.io/1qta

FreeBSD has similar issues when running tests for github.com/syncthing/syncthing/lib/fs

Note: The FreeBSD port does not appear to run tests when building. (Did not see anything in the Makefile for net/syncthing)

This might confirm that the issue is *BSD related, not just OpenBSD. Hope this helps.

**Comment by AudriusButkevicius** [comment]
So getting access to a bsd machine could help, but I am living out of a suitcase with no access to a proper pc for potentially one more month, so I can't help in the short term.

**Comment by vext01** [comment]
@AudriusButkevicius Are you back home? Just wondering if we could take a look at this soon. If so, I can get you a BSD machine up and  running.

Thanks

**Comment by AudriusButkevicius** [comment]
Not yet sadly... Moving in Sunday, but I suspect I won't have access to my PC till Christmas.

**Comment by AudriusButkevicius** [comment]
Hey, I am finally back online. Send me the details over email, I'll have a look once I've done other non-fun adult stuff.

**Comment by AudriusButkevicius** [comment]
This seems to pass on the VM I was given.

**Comment by vext01** [comment]
Hi,

I'm able to reproduce this one on my laptop. It seems sensitive to the machine we are running on. I remember not being able to repro on my desktop.

I suppose this is the same inotify issue @AudriusButkevicius ? If so, can we skip this test also?

```
--- FAIL: TestParentDeletion (0.02s)
    requests_test.go:606: remove /tmp/_request-182950755/foo: directory not empty
FAIL
FAIL    github.com/syncthing/syncthing/lib/model        58.983s
```

^ on master.

Thanks

**Comment by imsodin** [comment]
That test does not involve the filesystem watcher, so it should be unrelated. Should because I think the filesystem watcher is not deactivated and might thus influence the test scenario, potentially in ways to make it fail spuriously (haven't actually looked at the code, just remember explicit full scans being used before checking test conditions).

**Comment by vext01** [comment]
Should we re-open this then?

**Comment by AudriusButkevicius** [comment]
Only if you have a way to reproduce this and debug this. I don't want to have tickets lying around which are impossible to fix.

**Comment by vext01** [comment]
OK, I'm going to run this test in a loop on the VM and see if I can get it to fail.

**Comment by vext01** [comment]
Hrm, fwiw, I'm using go-1.11.4.

Here's more information about failures. There's actually three failing tests here, but let's look at `TestParentDeletion` first:
https://gist.github.com/vext01/8ab7c6c134ccda6f9a2480ddd10e8ba3

BTW, what's the command to run that test in isolation. The command I used earlier in the issue doesn't seem to work any more:
```
$ STRACE=model go test -v -run TestParentDeletion 
can't load package: package github.com/syncthing/syncthing: build constraints exclude all Go files in /home/edd/go/src/github.com/syncthing/syncthing
```

**Comment by imsodin** [comment]
You must be in the directory of the package that the test belongs to, here `lib/model` - you cannot run specific tests from the repository root.

Embarrassingly I didn't read the issue and thus haven't seen that you already discovered that `os.RemovAll` is the source. To check for simultaneous creation in the to be removed directory, you should add `fs` to the STTRACE env var.

**Comment by vext01** [comment]
Thanks imsodin.

I'm not sure how insightful this will be, but I ran the test in a loop until it failed. It's non-deterministic.

```
$ for i in {0..50}; do env STRACE="model fs" go test -v -run TestParentDeletion; done
...
=== RUN   TestParentDeletion
15:12:14 INFO: No stored folder metadata for "default": recalculating
15:12:14 INFO: Ready to synchronize "default" (default) (sendreceive)
15:12:14 INFO: Device GYRZZQB-IRNPV4Z-T7TC52W-EQYJ3TT-FDQW6MW-DFLMU42-SSSU6EM-FBK2VAY client is " " named "" at                                                                                                    
15:12:14 INFO: Completed initial scan of sendreceive folder "default" (default)
--- PASS: TestParentDeletion (0.59s)
PASS
ok      github.com/syncthing/syncthing/lib/model        0.724s
```

vs. 

```
=== RUN   TestParentDeletion
15:12:16 INFO: No stored folder metadata for "default": recalculating
15:12:16 INFO: Ready to synchronize "default" (default) (sendreceive)
15:12:16 INFO: Device GYRZZQB-IRNPV4Z-T7TC52W-EQYJ3TT-FDQW6MW-DFLMU42-SSSU6EM-FBK2VAY client is " " named "" at                                                                                                    
15:12:16 INFO: Completed initial scan of sendreceive folder "default" (default)
--- FAIL: TestParentDeletion (0.07s)
    requests_test.go:606: remove /tmp/_request-879856505/foo: directory not empty
FAIL
```

**Comment by vext01** [comment]
Here's a paste showing two consecutive runs of the test where the first fails and the second succeeds:

https://gist.github.com/vext01/c1f815f8fe0f78a2ae17aa8920d62cca

**Comment by imsodin** [comment]
Can you apply this change and try whether you can reproduce it again please.

```diff
diff --git a/lib/model/requests_test.go b/lib/model/requests_test.go
index c6d9e3d1f..e179bb8ae 100644
--- a/lib/model/requests_test.go
+++ b/lib/model/requests_test.go
@@ -601,6 +601,8 @@ func TestParentDeletion(t *testing.T) {
                t.Fatalf("timed out")
        }
 
+       m.ScanFolders()
+
        // Delete parent dir
        if err := testFs.RemoveAll(parent); err != nil {
                t.Fatal(err)
```

**Comment by vext01** [comment]
Sadly this didn't seem to solve the issue:
https://gist.github.com/vext01/f58fb7e402b875165b50215006b5d40a

**Comment by imsodin** [comment]
And with this?

```diff
diff --git a/lib/model/requests_test.go b/lib/model/requests_test.go
index c6d9e3d1f..622cd5f4c 100644
--- a/lib/model/requests_test.go
+++ b/lib/model/requests_test.go
@@ -601,6 +601,10 @@ func TestParentDeletion(t *testing.T) {
 		t.Fatalf("timed out")
 	}
 
+	if err := m.ScanFolder("default"); err != nil {
+		t.Fatal("Failed scanning:", err)
+	}
+
 	// Delete parent dir
 	if err := testFs.RemoveAll(parent); err != nil {
 		t.Fatal(err)
@@ -718,6 +722,7 @@ func setupModelWithConnection() (*Model, *fakeConnection, string, *config.Wrappe
 	cfg := defaultCfgWrapper.RawCopy()
 	cfg.Devices = append(cfg.Devices, config.NewDeviceConfiguration(device2, "device2"))
 	cfg.Folders[0] = config.NewFolderConfiguration(protocol.LocalDeviceID, "default", "default", fs.FilesystemTypeBasic, tmpDir)
+	cfg.Folders[0].FSWatcherEnabled = false
 	cfg.Folders[0].Devices = []config.FolderDeviceConfiguration{
 		{DeviceID: device1},
 		{DeviceID: device2},
```

**Comment by vext01** [comment]
Looks like this may have done the trick. I ran 1000 times with no failures.

I'll do another 3000 to be sure.

**Comment by imsodin** [comment]
@vext01 Can you hold off on that until I PR or remove the `ScanFolder` part? On second thoughts that should have no influence at all on whether the test passes or fails.

**Comment by vext01** [comment]
Re-running with only the second patch hunk :)

**Comment by vext01** [comment]
I ran the one test in a loop on the imsodin's branch for well over an hour, and saw no failure, so I think this change has fixed this test.

I still have failures on:
 * TestIssue2571
 * TestInternalScan
 * TestRequestVersioningSymlinkAttack
 * TestReadOnlyDir

So it hasn't resolved all of our problems.

I'll post more info on the above failures (in the appropriate places) once this issue is has been resolved.

Thanks :)
